src/components/modals/sleep-timer/sleep-timer.tsx

```diff
@@ -0,0 +1,127 @@
+import { useEffect, useState, useCallback } from 'react';
+import { Modal } from '@/components/modal';
+import { FaPlay, FaUndo } from 'react-icons/fa/index';
+import { useSoundStore } from '@/store';
+
+import { Button } from '@/components/generic/button';
+import { Timer } from '@/components/generic/timer';
+
+import styles from './sleep-timer.module.css';
+
+interface SleepTimerModalProps {
+  onClose: () => void;
+  show: boolean;
+}
+
+export function SleepTimerModal({ onClose, show }: SleepTimerModalProps) {
+  const [hours, setHours] = useState<string>('0');
+  const [minutes, setMinutes] = useState<string>('0');
+  const [running, setRunning] = useState(false);
+  const [timeLeft, setTimeLeft] = useState(0);
+  const [timerId, setTimerId] = useState<NodeJS.Timeout | undefined>(undefined);
+
+  const pause = useSoundStore(state => state.pause);
+
+  const calculateTotalSeconds = useCallback((): number => {
+    return (
+      (hours === '' ? 0 : parseInt(hours)) * 3600 +
+      (minutes === '' ? 0 : parseInt(minutes)) * 60
+    );
+  }, [minutes, hours]);
+
+  useEffect(() => {
+    setTimeLeft(calculateTotalSeconds());
+  }, [calculateTotalSeconds]);
+
+  // Handle multiple clicks on this. Only the latest click should be taken into account
+  const handleStart = () => {
+    if (timerId) clearInterval(timerId);
+
+    setTimeLeft(calculateTotalSeconds);
+    setRunning(true);
+
+    if (timeLeft > 0) {
+      const newTimerId = setInterval(() => {
+        setTimeLeft(prevTimeLeft => {
+          const newTimeLeft = prevTimeLeft - 1;
+          if (newTimeLeft <= 0) {
+            clearInterval(newTimerId);
+            pause();
+            setRunning(false);
+            return 0;
+          }
+          return newTimeLeft;
+        });
+      }, 1000);
+
+      setTimerId(newTimerId);
```

Comment:
It's better to use a ref instead of state for timerId.

Reply:
Going to check that ASAP. I will put everything in a second commit so you don't have to review everything a second time. I let you squash them if you want !